source/dofs/dof_handler.cc

```diff
@@ -2643,6 +2643,99 @@ DoFHandler<dim, spacedim>::distribute_mg_dofs()
 
 
 
+template <int dim, int spacedim>
+bool
+DoFHandler<dim, spacedim>::prepare_coarsening_and_refinement(
+  const unsigned int max_difference,
+  const unsigned int contains_fe_index) const
+{
+  Assert(max_difference > 0,
+         ExcMessage(
+           "This function does not serve any purpose for max_difference = 0."));
+  AssertIndexRange(contains_fe_index, fe_collection.size());
+
+  if (!hp_capability_enabled)
+    // nothing to do
+    return false;
+
+  // communication of future fe indices on ghost cells necessary.
+  // thus, currently only implemented for serial triangulations.
+  Assert((dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+            &(*tria)) == nullptr),
+         ExcNotImplemented());
+
+  //
+  // establish hierarchy
+  //
+  // - create bimap between hierarchy levels and FE indices
+  // - classify cells based on their level
+  const std::vector<unsigned int> fe_on_level =
+    fe_collection.get_hierarchy_sequence(contains_fe_index);
+
+  std::map<unsigned int, unsigned int> level_on_fe;
+  for (unsigned int i = 0; i < fe_on_level.size(); ++i)
+    level_on_fe[fe_on_level[i]] = i;
+
+  std::vector<std::list<active_cell_iterator>> cells_on_level(
+    fe_on_level.size());
+  if (fe_on_level.size() == fe_collection.size())
+    {
+      // all FE indices are part of hierarchy
+      for (const auto &cell : active_cell_iterators())
+        cells_on_level[level_on_fe[cell->future_fe_index()]].push_back(cell);
+    }
+  else
+    {
+      // some FE indices are part of hierarchy
+      typename decltype(level_on_fe)::iterator iter;
+      for (const auto &cell : active_cell_iterators())
+        if ((iter = level_on_fe.find(cell->future_fe_index())) !=
+            level_on_fe.end())
```

Comment:
we need C++17 with init-statement! https://en.cppreference.com/w/cpp/language/if